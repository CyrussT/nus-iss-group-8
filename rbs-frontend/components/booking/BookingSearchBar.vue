<script setup>
import { ref, defineEmits, onMounted, onBeforeUnmount } from 'vue';

const props = defineProps({
  loading: {
    type: Boolean,
    default: false
  },
  resourceTypeOptions: {
    type: Array,
    default: () => []
  },
  resourceNameOptions: {
    type: Array,
    default: () => []
  },
  locationOptions: {
    type: Array,
    default: () => []
  },
  optionsLoading: {
    type: Boolean,
    default: false
  }
});

const emit = defineEmits(['search', 'reset']);

const searchQuery = ref({
  resourceTypeId: "",
  resourceName: "",
  location: "",
  capacity: ""
});

const handleSearch = () => {
  emit('search', { ...searchQuery.value });
};

const resetSearchFields = () => {
  // Reset search form fields
  searchQuery.value = {
    resourceTypeId: "",
    resourceName: "",
    location: "",
    capacity: ""
  };
  console.log("Search fields have been reset");
};

const handleReset = () => {
  // Reset local search fields
  resetSearchFields();
  
  // Emit the reset event to parent
  emit('reset');
};

// Listen for external reset events 
const handleExternalReset = () => {
  resetSearchFields();
};

// Set up event listener for external reset
onMounted(() => {
  document.addEventListener('reset-search-fields', handleExternalReset);
});

// Clean up event listener
onBeforeUnmount(() => {
  document.removeEventListener('reset-search-fields', handleExternalReset);
});
</script>

<template>
  <UCard class="booking-search-bar p-4">
    <template #header>
      <div class="flex items-center">
        <UIcon name="i-heroicons-magnifying-glass" class="mr-2 text-gray-500 dark:text-gray-400" />
        <h3 class="text-lg font-medium dark:text-white">Find Available Resources</h3>
      </div>
    </template>

    <div class="mb-4 grid grid-cols-1 md:grid-cols-2 gap-4">
      <!-- Resource Type with UInputMenu -->
      <div>
        <label class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Resource Type</label>
        <UInputMenu
          v-model="searchQuery.resourceTypeId"
          :options="resourceTypeOptions"
          option-attribute="facilityType"
          value-attribute="facilityTypeId"
          placeholder="Type or select resource type"
          size="md"
          class="w-full"
          :loading="optionsLoading"
          clearable
        />
      </div>

      <!-- Resource Name with regular UInput - no autocomplete -->
      <div>
        <label class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Resource Name</label>
        <UInput
          v-model="searchQuery.resourceName"
          placeholder="Enter resource name"
          size="md"
          class="w-full"
          icon="i-heroicons-building-office"
          clearable
        />
      </div>

      <!-- Location with UInputMenu -->
      <div>
        <label class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Location</label>
        <UInputMenu
          v-model="searchQuery.location"
          :options="locationOptions"
          placeholder="Type or select location"
          size="md"
          class="w-full"
          :loading="optionsLoading"
          clearable
        />
      </div>

      <!-- Capacity -->
      <div>
        <label class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Minimum Capacity</label>
        <UInput 
          v-model="searchQuery.capacity" 
          type="number" 
          placeholder="Min. capacity needed"
          size="md"
          icon="i-heroicons-user-group"
        />
      </div>
    </div>

    <div class="flex justify-end gap-3">
      <UButton
        color="gray"
        variant="soft"
        icon="i-heroicons-arrow-path"
        @click="handleReset"
      >
        Reset
      </UButton>
      
      <UButton
        color="primary"
        icon="i-heroicons-magnifying-glass"
        @click="handleSearch"
        :loading="loading"
      >
        Search
      </UButton>
    </div>
  </UCard>
</template>

<style scoped>
/* Dark mode overrides for the search bar */
:deep(.dark .booking-search-bar) {
  background-color: #1e1e2d;
  border-color: #4a5568;
}

:deep(.dark .u-input) {
  background-color: #2d3748;
  border-color: #4a5568;
  color: #e2e8f0;
}

:deep(.dark .u-input::placeholder) {
  color: #a0aec0;
}

:deep(.dark .u-input-icon) {
  color: #a0aec0;
}

:deep(.dark .u-input-menu) {
  background-color: #2d3748;
  border-color: #4a5568;
}

:deep(.dark .u-input-menu__item) {
  color: #e2e8f0;
}

:deep(.dark .u-input-menu__item:hover) {
  background-color: #4a5568;
}

:deep(.dark .u-input-menu__item--selected) {
  background-color: #553c9a;
}

:deep(.dark .u-button--soft.u-button--gray) {
  background-color: #2d3748;
  color: #e2e8f0;
}

:deep(.dark .u-button--soft.u-button--gray:hover) {
  background-color: #4a5568;
}

:deep(.dark input[type="number"]) {
  color-scheme: dark;
}
</style>