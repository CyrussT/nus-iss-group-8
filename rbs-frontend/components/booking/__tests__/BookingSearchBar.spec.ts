import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import BookingSearchBar from '../BookingSearchBar.vue';

describe('BookingSearchBar', () => {
  it('renders correctly', () => {
    const wrapper = mount(BookingSearchBar, {
      global: {
        stubs: {
          UButton: true,
          UIcon: true
        }
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
