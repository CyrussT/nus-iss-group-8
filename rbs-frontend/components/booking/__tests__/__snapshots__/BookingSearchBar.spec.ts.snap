// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookingSearchBar > renders correctly 1`] = `
"<div data-v-c950d2e1="" class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900 booking-search-bar p-4">
  <div class="px-4 py-5 sm:px-6">
    <div data-v-c950d2e1="" class="flex items-center"><span data-v-c950d2e1="" class="iconify i-heroicons:magnifying-glass mr-2 text-gray-500 dark:text-gray-400" aria-hidden="true"></span>
      <h3 data-v-c950d2e1="" class="text-lg font-medium dark:text-white">Find Available Resources</h3>
    </div>
  </div>
  <div class="px-4 py-5 sm:p-6">
    <div data-v-c950d2e1="" class="mb-4 grid grid-cols-1 md:grid-cols-2 gap-4">
      <!-- Resource Type with UInputMenu -->
      <div data-v-c950d2e1=""><label data-v-c950d2e1="" class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Resource Type</label>
        <div class="relative w-full" data-headlessui-state="">
          <div class="flex items-center w-full"><input value="" id="headlessui-combobox-input-v-0" aria-expanded="false" aria-autocomplete="list" role="combobox" type="text" tabindex="0" data-headlessui-state="" placeholder="Type or select resource type" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-3 py-2 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 pe-10" autocomplete="off" clearable="">
            <!--v-if--><button id="headlessui-combobox-button-v-1" type="button" tabindex="-1" aria-haspopup="listbox" aria-expanded="false" data-headlessui-state="" class="absolute inset-y-0 end-0 flex items-center px-3">
              <u-icon-stub name="i-heroicons-chevron-down-20-solid" class="flex-shrink-0 text-gray-400 dark:text-gray-500 h-5 w-5"></u-icon-stub>
            </button>
          </div>
          <!--v-if-->
        </div>
      </div><!-- Resource Name with regular UInput - no autocomplete -->
      <div data-v-c950d2e1=""><label data-v-c950d2e1="" class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Resource Name</label>
        <div data-v-c950d2e1="" class="relative w-full"><input type="text" placeholder="Enter resource name" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-3 py-2 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 ps-10" clearable="" value=""><span class="absolute inset-y-0 start-0 flex items-center pointer-events-none px-3"><u-icon-stub name="i-heroicons-building-office" class="flex-shrink-0 text-gray-400 dark:text-gray-500 h-5 w-5"></u-icon-stub></span>
          <!--v-if-->
        </div>
      </div><!-- Location with UInputMenu -->
      <div data-v-c950d2e1=""><label data-v-c950d2e1="" class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Location</label>
        <div class="relative w-full" data-headlessui-state="">
          <div class="flex items-center w-full"><input value="" id="headlessui-combobox-input-v-2" aria-expanded="false" aria-autocomplete="list" role="combobox" type="text" tabindex="0" data-headlessui-state="" placeholder="Type or select location" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-3 py-2 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 pe-10" autocomplete="off" clearable="">
            <!--v-if--><button id="headlessui-combobox-button-v-3" type="button" tabindex="-1" aria-haspopup="listbox" aria-expanded="false" data-headlessui-state="" class="absolute inset-y-0 end-0 flex items-center px-3">
              <u-icon-stub name="i-heroicons-chevron-down-20-solid" class="flex-shrink-0 text-gray-400 dark:text-gray-500 h-5 w-5"></u-icon-stub>
            </button>
          </div>
          <!--v-if-->
        </div>
      </div><!-- Capacity -->
      <div data-v-c950d2e1=""><label data-v-c950d2e1="" class="block text-sm font-medium text-gray-700 dark:text-gray-300 mb-1">Minimum Capacity</label>
        <div data-v-c950d2e1="" class="relative"><input type="number" placeholder="Min. capacity needed" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-3 py-2 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 ps-10" value=""><span class="absolute inset-y-0 start-0 flex items-center pointer-events-none px-3"><u-icon-stub name="i-heroicons-user-group" class="flex-shrink-0 text-gray-400 dark:text-gray-500 h-5 w-5"></u-icon-stub></span>
          <!--v-if-->
        </div>
      </div>
    </div>
    <div data-v-c950d2e1="" class="flex justify-end gap-3"><button data-v-c950d2e1="" type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 px-2.5 py-1.5 text-gray-500 dark:text-gray-400 bg-gray-50 hover:bg-gray-100 disabled:bg-gray-50 aria-disabled:bg-gray-50 dark:bg-gray-950 dark:hover:bg-gray-900 dark:disabled:bg-gray-950 dark:aria-disabled:bg-gray-950 focus-visible:ring-2 focus-visible:ring-inset focus-visible:ring-gray-500 dark:focus-visible:ring-gray-400 inline-flex items-center">
        <u-icon-stub name="i-heroicons-arrow-path" class="flex-shrink-0 h-5 w-5" aria-hidden="true"></u-icon-stub> Reset
        <!--v-if-->
      </button><button data-v-c950d2e1="" type="button" class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 px-2.5 py-1.5 shadow-sm text-white dark:text-gray-900 bg-primary-500 hover:bg-primary-600 disabled:bg-primary-500 aria-disabled:bg-primary-500 dark:bg-primary-400 dark:hover:bg-primary-500 dark:disabled:bg-primary-400 dark:aria-disabled:bg-primary-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-primary-500 dark:focus-visible:outline-primary-400 inline-flex items-center">
        <u-icon-stub name="i-heroicons-magnifying-glass" class="flex-shrink-0 h-5 w-5" aria-hidden="true"></u-icon-stub> Search
        <!--v-if-->
      </button></div>
  </div>
  <!--v-if-->
</div>"
`;
